File: frontend/src/types/tracking.ts
import type { BoundingBox } from "@/types/detection";

export type TrackingSourceType = "webcam" | "video_file";

export interface TrackedObject {
  track_id: number;
  label: string;
  confidence: number;
  box: BoundingBox;
}

export interface TrackingStatusResponse {
  tracking: boolean;
  source: string | null;
  active_objects: number;
  fps: number | null;
}

export interface TrackingStateResponse {
  tracking: boolean;
  active_objects: number;
  fps: number | null;
  tracks: TrackedObject[];
}

/** Frontend-only UI state machine, layered on top of the backend's tracking boolean. */
export type TrackingUIStatus =
  | "idle"
  | "starting"
  | "tracking"
  | "stopping"
  | "error";

File: backend/app/schemas/tracking.py
"""Pydantic schemas for the real-time detection-and-tracking API."""

from enum import Enum

from pydantic import BaseModel, Field

from app.schemas.detection import BoundingBox


class TrackingSourceType(str, Enum):
    """Where tracking frames come from."""

    WEBCAM = "webcam"
    VIDEO_FILE = "video_file"


class TrackedObjectSchema(BaseModel):
    """A single tracked object, exposed over the JSON polling endpoint."""

    track_id: int = Field(..., description="Persistent ID; stable across frames.")
    label: str = Field(..., description="Predicted class label.")
    confidence: float = Field(..., ge=0.0, le=1.0)
    box: BoundingBox


class TrackingStatusResponse(BaseModel):
    """Returned by /tracking/start and /tracking/stop."""

    tracking: bool
    source: str | None = None
    active_objects: int
    fps: float | None = None


class TrackingStateResponse(BaseModel):
    """Returned by GET /tracking/state for the live sidebar UI to poll."""

    tracking: bool
    active_objects: int
    fps: float | None = None
    tracks: list[TrackedObjectSchema]